source/extensions/filters/http/common/aws/credentials_provider_impl.cc

```diff
@@ -0,0 +1,223 @@
+  if (needsRefresh()) {
+    refresh();
+  }
+}
+
+bool InstanceProfileCredentialsProvider::needsRefresh() {
+  return api_.timeSource().systemTime() - last_updated_ > REFRESH_INTERVAL;
+}
+
+void InstanceProfileCredentialsProvider::refresh() {
+  ENVOY_LOG(debug, "Getting AWS credentials from the instance metadata");
+
+  // First discover the Role of this instance
+  const auto instance_role_string =
+      metadata_fetcher_(EC2_METADATA_HOST, SECURITY_CREDENTIALS_PATH, absl::nullopt);
+  if (!instance_role_string) {
+    ENVOY_LOG(error, "Could not retrieve credentials listing from the instance metadata");
+    return;
+  }
+
+  const auto instance_role_list =
+      StringUtil::splitToken(StringUtil::trim(instance_role_string.value()), "\n");
+  if (instance_role_list.empty()) {
+    ENVOY_LOG(error, "No AWS credentials were found in the instance metadata");
+    return;
+  }
+  ENVOY_LOG(debug, "AWS credentials list:\n{}", instance_role_string.value());
+
+  // Only one Role can be associated with an instance:
+  // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html
+  const auto credential_path =
+      std::string(SECURITY_CREDENTIALS_PATH) + "/" +
+      std::string(instance_role_list[0].data(), instance_role_list[0].size());
+  ENVOY_LOG(debug, "AWS credentials path: {}", credential_path);
+
+  // Then fetch and parse the credentials
+  const auto credential_document =
+      metadata_fetcher_(EC2_METADATA_HOST, credential_path, absl::nullopt);
+  if (!credential_document) {
+    ENVOY_LOG(error, "Could not load AWS credentials document from the instance metadata");
+    return;
+  }
+
+  Json::ObjectSharedPtr document_json;
+  try {
+    document_json = Json::Factory::loadFromString(credential_document.value());
+  } catch (EnvoyException& e) {
+    ENVOY_LOG(error, "Could not parse AWS credentials document: {}", e.what());
+    return;
+  }
+
+  const auto access_key_id = document_json->getString(ACCESS_KEY_ID, "");
+  const auto secret_access_key = document_json->getString(SECRET_ACCESS_KEY, "");
+  const auto session_token = document_json->getString(TOKEN, "");
+
+  ENVOY_LOG(debug, "Found following AWS credentials in the instance metadata: {}={}, {}={}, {}={}",
+            AWS_ACCESS_KEY_ID, access_key_id, AWS_SECRET_ACCESS_KEY,
+            secret_access_key.empty() ? "" : "*****", AWS_SESSION_TOKEN,
+            session_token.empty() ? "" : "*****");
+
+  cached_credentials_ = Credentials::fromString(access_key_id, secret_access_key, session_token);
+  last_updated_ = api_.timeSource().systemTime();
+}
+
+bool TaskRoleCredentialsProvider::needsRefresh() {
+  const auto now = api_.timeSource().systemTime();
+  return (now - last_updated_ > REFRESH_INTERVAL) ||
+         (expiration_time_ - now < REFRESH_GRACE_PERIOD);
+}
+
+void TaskRoleCredentialsProvider::refresh() {
+  ENVOY_LOG(debug, "Getting AWS credentials from the task role at URI: {}", credential_uri_);
+
+  absl::string_view host_view;
```

Comment:
`absl::string_view host, path;` for brevity?

Reply:
fixed